software/tracksight/frontend/components/Legend.tsx

```diff
@@ -0,0 +1,57 @@
+import ChipLegendRenderer from "@/components/ChipLegendItem";
+import LabelLegendItem from "@/components/LabelLegendItem";
+import React, { useMemo } from "react";
+
+type LegendThemes = "chip" | "label";
+
+type LegendProps = {
+  theme: LegendThemes;
+
+  // TODO(evan): Implement editable legends
+  onAddSignal?: (signal: string) => void;
+  onRemoveSignal?: (signal: string) => void;
+
+  signals: string[];
+  colorPalette: string[];
+};
+
+type LegendItemRendererProps = {
+  content: string;
+  color: string;
+
+  removable?: boolean;
+  onRemoved?: () => {};
+};
+
+export type LegendItemRenderer = React.FC<LegendItemRendererProps>;
+
+const LEGEND_ITEM_RENDERERS: Record<LegendThemes, LegendItemRenderer> = {
+  chip: ChipLegendRenderer,
+  label: LabelLegendItem,
+};
+
+const Legend: React.FC<LegendProps> = (props) => {
+  const {
+    theme,
+
+    signals,
+    colorPalette,
+  } = props;
+
+  const ItemRenderer = useMemo(() => LEGEND_ITEM_RENDERERS[theme], [theme]);
+
+  return (
+    <div
+      className="grid w-full max-w-full gap-4 px-8"
+      style={{
+        gridTemplateColumns: "repeat(auto-fit, 250px)",
+      }}
+    >
+      {signals.map((signal, i) => (
+        <ItemRenderer key={signal} content={signal} color={colorPalette[i] || "#000000"} />
```

Comment:
just branch here, it's faster and does not consume react state memory

Reply:
like honestly don't even branch here just have two seperate components like this is doing too much

Reply:
Yeah again probably a bad habit, I just drop this in whenever there's a conditional on a renderer. Ill switch this pattern out.